File: projects/grind-75/python/tree/binary_tree_right_side_view.py
from collections import deque
from typing import List, Optional
from tree.common import TreeNode


class Solution:
    @classmethod
    def rightSideViewDfs(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node=root, i=0):
            if not node:
                return
            if i == len(res):
                res.append(node.val)
            dfs(node.right, i + 1)
            dfs(node.left, i + 1)

        dfs()
        return res

    @classmethod
    def rightSideViewBfs(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if not root:
            return res
        q = deque([root])
        while q:
            res.append(q[-1].val)
            for i in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return res

File: projects/grind-75/typescript/tree/binary-tree-right-side-view.ts
import { TreeNode } from './common';

export default function rightSideView(root: TreeNode | null): number[] {
  const result: number[] = [];
  if (!root) return result;

  const queue: TreeNode[] = [root];
  let start = 0;

  while (start < queue.length) {
    const levelSize = queue.length - start;
    result.push(queue[queue.length - 1].val);

    for (let i = 0; i < levelSize; i++) {
      const node = queue[start];

      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
      start++;
    }
  }

  return result;
}
